lib/srv/reexec.go

```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
+
+		listener, err := newListener(ctx, "unix", socketPath)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer listener.Close()
+
+		listenerFD, err := getListenerFile(listener)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// Update the agent forwarding socket with more restrictive permissions.
+		if err = os.Chmod(socketPath, teleport.FileMaskOwnerOnly); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		return listenerFD, trace.Wrap(err)
+
+	case networking.NetworkingOperationListenX11:
+		listener, display, err := newX11Listener(ctx, req.X11Request.DisplayOffset, req.X11Request.MaxDisplay)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer listener.Close()
+
+		// Setup the user's local xauth file to interface with the local x11 listener.
+		removeCmd := x11.NewXAuthCommand(ctx, "")
+		if err := removeCmd.RemoveEntries(display); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		addCmd := x11.NewXAuthCommand(ctx, "")
+		if err := addCmd.AddEntry(x11.XAuthEntry{
+			Display: display,
+			Proto:   req.X11Request.AuthProtocol,
+			Cookie:  req.X11Request.AuthCookie,
+		}); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		listenerFD, err := getListenerFile(listener)
+		return listenerFD, trace.Wrap(err)
+
+	default:
+		return nil, trace.BadParameter("unsupported networking operation %q", req.Operation)
+	}
 }
 
-// RunRemoteForward reads in the command to run from the parent process (over a
-// pipe) then listens for port forwarding.
-func RunRemoteForward() (errw io.Writer, code int, err error) {
-	errw, code, err = runForward(handleRemotePortForward)
-	return errw, code, trace.Wrap(err)
+func getListenerFile(listener net.Listener) (*os.File, error) {
+	switch l := listener.(type) {
+	case *net.UnixListener:
+		l.SetUnlinkOnClose(false)
+		listenerFD, err := l.File()
+		return listenerFD, trace.Wrap(err)
+	case *net.TCPListener:
+		listenerFD, err := l.File()
+		return listenerFD, trace.Wrap(err)
+	default:
+		return nil, trace.Errorf("expected listener to be of type *net.UnixListener or *net.TCPListener, but was %T", l)
+	}
+}
+
+func getConnFile(conn net.Conn) (*os.File, error) {
+	switch c := conn.(type) {
+	case *net.UnixConn:
+		connFD, err := c.File()
+		return connFD, trace.Wrap(err)
+	case *net.TCPConn:
+		connFD, err := c.File()
+		return connFD, trace.Wrap(err)
+	default:
+		return nil, trace.Errorf("expected connection to be of type *net.UnixConn or *net.TCPConn, but was %T", conn)
+	}
+}
+
+// newListener creates a new network listener with address reuse disabled.
+func newListener(ctx context.Context, network string, addr string) (net.Listener, error) {
+	lc := net.ListenConfig{
+		Control: func(network, addr string, conn syscall.RawConn) error {
+			var err error
+			err2 := conn.Control(func(descriptor uintptr) {
+				// Disable address reuse to prevent socket replacement.
+				err = syscall.SetsockoptInt(int(descriptor), syscall.SOL_SOCKET, syscall.SO_REUSEADDR, 0)
```

Comment:
Is SO_REUSEADDR ever enabled by default?

Reply:
Re-tested this and it seems it's only enabled by default for tcp listeners.

Reply:
Is this necessary? If so, for what? `SO_REUSEADDR` allows listeners to be opened on top of sockets in `TIME_WAIT` state, opening multiple listeners on the same bind address at the same time is done with `SO_REUSEPORT`.

Reply:
I'm not sure, I've preserved this from the remote port forwarding implementation. @atburke is this check necessary for some purpose or can it be removed?

Reply:
I believe `SO_REUSEADDR` was added here to prevent some malicious process from being able to hijack the address. You'd probably be able to get a more complete answer from @jentfoo.